src/utils.rs

```diff
@@ -273,15 +273,43 @@ pub(crate) fn contains_skip(attrs: &[Attribute]) -> bool {
 #[inline]
 pub(crate) fn semicolon_for_expr(context: &RewriteContext<'_>, expr: &ast::Expr) -> bool {
     // Never try to insert semicolons on expressions when we're inside
-    // a macro definition - this can prevent the macro from compiling
+    // a macro definition - this  can prevent the macro from compiling
     // when used in expression position
     if context.is_macro_def {
         return false;
     }
 
     match expr.kind {
         ast::ExprKind::Ret(..) | ast::ExprKind::Continue(..) | ast::ExprKind::Break(..) => {
-            context.config.trailing_semicolon()
+            match context.config.trailing_semicolon() {
+                TrailingSemicolon::Always => true,
+                TrailingSemicolon::Never | TrailingSemicolon::Preserve => false,
+            }
+        }
+        _ => false,
+    }
+}
+
+/// Previously, we used to have `trailing_semicolon = always` enabled, and due to
+/// a bug between `format_stmt` and `format_expr`, we used to subtract the size of
+/// `;` *TWICE* from the shape. This means that an expr that would fit onto a line
+/// of, e.g. 99 (limit 100) after subtracting one for the semicolon would still be
+/// wrapped.
+///
+/// This function reimplements the old heuristic of double counting the "phantom"
+/// semicolon that should have already been accounted for, to not break existing
+/// formatting with the `trailing_semicolon = preserve` behavior.
```

Comment:
Would you mind showing an example where things are wrapping early, maybe even adding a `version=One` test case if we go the route of fixing things for `version=Two`.

Reply:
```
fn main() {
    return hellooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo(arg);
}
```

The return statement line is exactly 100 chars long. When formatting the expression, we first subtract one from the shape here:
https://github.com/rust-lang/rustfmt/blob/728939191e4218e2c1296c7ba3eb36590cbcb9bd/src/stmt.rs#L119-L128

(See `let shape = shape.sub_width(suffix.len())?; `)

And then once again here:
https://github.com/rust-lang/rustfmt/blob/728939191e4218e2c1296c7ba3eb36590cbcb9bd/src/expr.rs#L67-L71

(in `format_expr`).

In practice, for example, we need to format `return helloo...o(arg)` which is an expression that is *95* characters long (100 - 4 identation - 1 semicolon) but after subtracting from the `shape` twice, we only have have a budget of *94* characters, even though the whole expression should fit in the line.

I can add it as a `version=One` test.

Reply:
For the record, i discovered this because it happens twice in practice in the rustfmt codebase, so the self-formatting test failed.

Reply:
Yeah, we use `version=Two` within rustfmt as a way to dogfood unstable formatting (at least that's my assumption)